src/compiler/checker.ts

```diff
@@ -6568,6 +6616,394 @@ module ts {
+                }
+                else {
+                    // Resolution failed
+                    return anyType;
+                }
+            }
+        }
+
+        /// Given a JSX attribute, returns the symbol for the corresponds property
+        /// of the element attributes type. Will return unknownSymbol for attributes
+        /// that have no matching element attributes type property.
+        function getJsxAttributePropertySymbol(attrib: JsxAttribute): Symbol {
+            let attributesType = getJsxElementAttributesType(<JsxOpeningElement>attrib.parent);
+            let prop = getPropertyOfType(attributesType, attrib.name.text);
+            return prop || unknownSymbol;
+        }
+
+        function getJsxGlobalElementClassType(): Type {
+            var jsxNS = getGlobalSymbol(JsxNames.JSX, SymbolFlags.Namespace, /*diagnosticMessage*/ undefined);
+            if (jsxNS) {
+                let sym = getSymbol(jsxNS.exports, JsxNames.ElementClass, SymbolFlags.Type);
+                let elemClassType = sym && getDeclaredTypeOfSymbol(sym);
+                return elemClassType;
+            }
+            else {
+                return undefined;
+            }
+        }
+
+        /// Returns all the properties of the Jsx.IntrinsicElements interface
+        function getJsxIntrinsicTagNames(): Symbol[]{
+            let intrinsics = getJsxIntrinsicElementsType();
+            return intrinsics ? getPropertiesOfType(intrinsics) : [];
+        }
+
+        function checkJsxOpeningElement(node: JsxOpeningElement) {
+            var targetAttributesType = getJsxElementAttributesType(node);
+
+            var nameTable: Map<boolean> = {};
+            // Process this array in right-to-left order so we know which
+            // attributes (mostly from spreads) are being overwritten and
+            // thus should have their types ignored
+            for (var i = node.attributes.length - 1; i >= 0; i--) {
+                if (node.attributes[i].kind === SyntaxKind.JsxAttribute) {
+                    checkJsxAttribute(<JsxAttribute>(node.attributes[i]), targetAttributesType, nameTable);
+                }
+                else if (node.attributes[i].kind === SyntaxKind.JsxSpreadAttribute) {
+                    checkJsxSpreadAttribute(<JsxSpreadAttribute>(node.attributes[i]), targetAttributesType, nameTable);
+                }
+                else {
+                    throw new Error('Unknown JSX attribute kind');
+                }
+            }
+
+            // Check that all required properties have been provided
+            if (targetAttributesType) {
+                let targetProperties = getPropertiesOfType(targetAttributesType);
+                for (var i = 0; i < targetProperties.length; i++) {
+                    if (!(targetProperties[i].flags & SymbolFlags.Optional) &&
+                        nameTable[targetProperties[i].name] === undefined) {

```

Comment:
Just curious, is it possible to have an attribute name that is the empty string?